{% extends "layout.html" %}

{% block body %}
<style>
    .containery {
        display: grid;
        padding: 20px;
    }
    .button-containery {
        max-width: 2000px; /* Optional: Limit the width */
        max-height: 1000px; /* Set the maximum height */
        overflow-y: auto; /* Enable vertical scrolling for long lists */
    }
    .plot-container {
        padding: 20px;
    }
    .data-container {
        padding: 20px;
    }
    .gene-buttons {
        display: flex;
        flex-wrap: wrap;
        gap: 1px; /* Space between buttons */
        justify-content: left; /* Center-align buttons */
        margin-top: 10px;
    }

    .gene-button {
        background-color: steelblue; /* Blue background */
        color: white; /* White text */
        border: none;
        border-radius: 5px;
        padding: 5px 10px;
        font-size: 12px;
        cursor: pointer;
        transition: background-color 0.3s ease;
    }

    .gene-button:hover {
        background-color: #1678e0; 
    }
</style>

<div id="containery" class="containery">
    <div id="button-containery" class="button-containery">
        <div>
            <button class="gene-button"><a href="{{ url_for('cov_bp.show_blacklisted_regions', group=smp_grp)}}">Blacklisted Regions</a></button>
        </div>
        Using the following genelists:
        {% for list in genelists %}
            {{list}}
        {% endfor %}
        <form action="" method="POST" name="form" class="no-spin">
            <form action="{{ url_for('cov_bp.get_cov', sample_id=sample.name) }}" method="POST">
                <button type="submit" class="gene-button" name="depth_cutoff" value="100">100X</button>
                <button type="submit" class="gene-button" name="depth_cutoff" value="500">500X</button>
                <button type="submit" class="gene-button" name="depth_cutoff" value="1000">1000X</button>
            </form>
        </form>
    
        <div class="gene-buttons">
              <!-- Low Coverage Regions -->
            <div id="lowcovlist_div"  clear:left; class="bg-white shadow-md rounded-lg p-6 mb-6">
                <h2 class="text-xl font-semibold text-gray-800 mb-4">Low Coverage Regions</h2>
                <div class="overflow-auto">
                    <table class="min-w-full mt-4 table-auto text-xs" id="sortable-table">
                        <thead>
                            <tr class="bg-gradient-to-r from-yellow-300 to-green-300 text-black">
                                <th class="py-2 px-4 border-b border-gray-300 text-center uppercase font-normal cursor-pointer" data-column="0" data-type="text">Chromosomal region</th>
                                <th class="py-2 px-4 border-b border-gray-300 text-center uppercase font-normal cursor-pointer" data-column="1" data-type="number">Size (bp)</th>
                                <th class="py-2 px-4 border-b border-gray-300 text-center uppercase font-normal cursor-pointer" data-column="2" data-type="number">Avg depth (x)</th>
                                <th class="py-2 px-4 border-b border-gray-300 text-center uppercase font-normal cursor-pointer" data-column="3" data-type="text">Exon</th>
                                <th class="py-2 px-4 border-b border-gray-300 text-center uppercase font-normal cursor-pointer" data-column="4" data-type="text">Gene</th>
                            </tr>
                        </thead>
                        <tbody>
                            {% for gene in cov_table %}
                            {% for CDS in cov_table[gene] %}
                            {% set exon = cov_table[gene][CDS] %}
                            {% if '_' in CDS %}
                            {% set CDS = "NCP" %}
                            {% endif %}
                            <tr class="hover:bg-gray-100">
                                <td class="py-2 px-4 border-b border-gray-300">{{exon.chr}}:{{exon.start}}-{{exon.end}}</td>
                                <td class="py-2 px-4 border-b border-gray-300">{{exon.end | int - exon.start | int}}</td>
                                <td class="py-2 px-4 border-b border-gray-300">{{exon.cov | float | round() }}</td>
                                <td class="py-2 px-4 border-b border-gray-300">{{CDS}}</td>
                                <td class="py-2 px-4 border-b border-gray-300"><button onclick="plotGene('{{gene}}')">{{gene}}</button></td>
                            </tr>
                            {% endfor %}
                            {% endfor %}
                        </tbody>
                    </table>
                </div>
            </div>
        </div>
    </div>
    <div class="plot-container" id="plot-container"></div>
    <div class="data-container" id="data-container"></div>
</div>
<script src="https://d3js.org/d3.v7.min.js"></script>

<script>
    function plotGene(gene) {
        // Parse the data from Jinja2
        const geneData = {{ coverage | tojson | safe }};
        const depthCutoff = {{ cov_cutoff }}
        // Get the specific gene data dynamically
        const selectedGene = geneData.genes[gene];
        if (!selectedGene) {
            console.error(`Gene ${gene} not found in the data.`);
            return;
        }

        // Remove any existing plot
        d3.select("#plot-container").select("#gene-plot-container").remove();
        
        // Create a new container for the plot
        const plotContainer = d3.select("#plot-container")
            .append("div")
            .attr("id", "gene-plot-container")
            .style("margin", "20px 0");

        // Add a title or information
        plotContainer.append("h3").text(`Gene: ${gene} @ ${depthCutoff}X `)
            .append("button")
            .style("color", "red")
            .text("Blacklist")
            .on("click", () => blacklistRegion(gene,"",""));

        // Create the SVG for the plot
        const svg = plotContainer.append("svg");

        // Calculate dynamic width based on gene length
        
        const geneStart = selectedGene.transcript.start
        const geneEnd = selectedGene.transcript.end
        const geneLength = geneEnd - geneStart;
        const baseWidthPerBp = 0.01; // Scale factor: 0.1px per base pair
        const svgWidth = Math.min(Math.max(geneLength * baseWidthPerBp, 2000), 2000); // always scale to 2000 BP
        const svgHeight = 200;
        const margin = { top: 20, right: 200, bottom: 20, left: 200 };
        const plotHeight = svgHeight - margin.top - margin.bottom;

        svg.attr("width", svgWidth)
            .attr("height", svgHeight);

        // Dynamic scale based on gene size
        const xScale = d3.scaleLinear()
            .domain([geneStart, geneEnd])
            .range([margin.left, svgWidth - margin.right]);

        const geneY = plotHeight / 2;

        // Draw the gene line
        svg.append("line")
            .style("stroke", "black")
            .style("stroke-width", 1)
            .attr("x1", xScale(selectedGene.transcript.start))
            .attr("y1", geneY)
            .attr("x2", xScale(selectedGene.transcript.end))
            .attr("y2", geneY);


        // Draw exons as rectangles
        svg.selectAll("exon")
            .data(selectedGene.exons)
            .enter()
            .append("rect")
            .style("fill", "lightgrey")
            .attr("x", d => xScale(d.start))
            .attr("y", geneY - 10)
            .attr("width", d => xScale(d.end) - xScale(d.start))
            .attr("height", 20)
            .append("title") // Tooltip
            .text(d => `${d.chr}:${d.start}-${d.end} (Exon ${d.nbr}), Cov: ${Number(d.cov).toFixed(2)}`);

        // Draw coding regions as rectangles
        svg.selectAll("cds")
            .data(selectedGene.CDS)
            .enter()
            .append("rect")
            .attr("fill", d => {
                    // Color based on coverage cutoff
                    return d.cov < depthCutoff ? "red" : "steelblue"; // Red if cov < 500, green otherwise
                })
            .attr("x", d => xScale(d.start))
            .attr("y", geneY - 10)
            .attr("width", d => xScale(d.end) - xScale(d.start))
            .attr("height", 20)
            .append("title") // Tooltip
            .text(d => `${d.chr}:${d.start}-${d.end} (Exon ${d.nbr}), Cov: ${Number(d.cov).toFixed(2)}`);
        
        svg.selectAll("probes")
            .data(selectedGene.probes)
            .enter()
            .append("rect")
            .attr("fill", d => {
                    // Color based on coverage cutoff
                    return d.cov < depthCutoff ? "red" : "steelblue"; // Red if cov < 500, green otherwise
                })
            .attr("x", d => xScale(d.start))
            .attr("y", geneY - 35)
            .attr("width", d => xScale(d.end) - xScale(d.start))
            .attr("height", 20)
            .append("title") // Tooltip
            .text(d => `${d.chr}:${d.start}-${d.end}, Cov: ${Number(d.cov).toFixed(2)}X`);
        
        // Remove old data before adding a new gene to the container
        d3.select("#data-container").select("#gene-data-container").remove();
        
        // Create a new container for the data
        const dataContainer = d3.select("#data-container")
            .append("div")
            .attr("id", "gene-data-container")
            .style("margin", "20px 0");
        


        // CODING SEQUENCE
        const lowCoverageExons = selectedGene.CDS.filter(d => d.cov < depthCutoff);
        if (lowCoverageExons.length > 0) {
            dataContainer.append("h4")
            .style("font-size", "18px")
            .style("font-weight", "bold")
            .text("Exons not meeting criteria:");
            // Create CDS table
            const table = dataContainer.append("table")
                .style("border-collapse", "collapse")
                .style("width", "auto") // Set table width to adjust to its content
                .style("margin-left", "0") // Align the table to the left
                .style("margin-top", "20px"); // Optional: Add some spacing at the top

            // Add table headers
            const thead = table.append("thead");
            thead.append("tr")
                .selectAll("th")
                .data(["Exon", "Coordinates", "Coverage", "Actions"])
                .enter()
                .append("th")
                .style("padding", "4px")
                .style("text-align", "left")
                .text(d => d);

            // Add table body
            const tbody = table.append("tbody");

            // Populate table rows with data
        
            lowCoverageExons.forEach(exon => {
                const row = tbody.append("tr");
                const coord = `${exon.chr}:${exon.start}-${exon.end}`
                // Add data cells
                row.append("td")
                    .style("padding", "4px")
                    .text(`${exon.nbr}`);
                row.append("td")
                    .style("padding", "4px")
                    .text(`${coord}`);
                row.append("td")
                    .style("padding", "4px")
                    .text(`${Number(exon.cov).toFixed(2)}X`);

                // Add the "Blacklist" button
                row.append("td")
                    .style("padding", "4px")
                    .style("color", "red")
                    .append("button")
                    .text("Blacklist")
                    .on("click", () => blacklistRegion(gene,coord,"CDS"));
            });
        }

        // PROBES 
        const lowCoverageProbes = selectedGene.probes.filter(d => d.cov < depthCutoff);
        // if probe data
        if (lowCoverageProbes.length > 0) {
            dataContainer.append("br");
            dataContainer.append("br");
            dataContainer.append("h4")
                .style("font-size", "18px")
                .style("font-weight", "bold")
                .text("Probes not meeting criteria:");
        
            // Create probe table
            const probe_table = dataContainer.append("table")
                .style("border-collapse", "collapse")
                .style("width", "auto") // Set table width to adjust to its content
                .style("margin-left", "0") // Align the table to the left
                .style("margin-top", "20px"); // Optional: Add some spacing at the top
        
            // Add table headers
            const thead_probe = probe_table.append("thead");
            thead_probe.append("tr")
                .selectAll("th")
                .data(["Coordinates", "Coverage", "Actions"])
                .enter()
                .append("th")
                .style("padding", "4px")
                .style("text-align", "left")
                .text(d => d);

            // Add table body
            const tbody_probe = probe_table.append("tbody");

        
            lowCoverageProbes.forEach(probe => {
                const row = tbody_probe.append("tr");
                const coord = `${probe.chr}:${probe.start}-${probe.end}`
                // Add data cells
                row.append("td")
                    .style("padding", "4px")
                    .text(`${coord}`);
                row.append("td")
                    .style("padding", "4px")
                    .text(`${Number(probe.cov).toFixed(2)}X`);

                // Add the "Blacklist" button
                row.append("td")
                    .style("padding", "4px")
                    .style("color", "red")
                    .append("button")
                    .text("Blacklist")
                    .on("click", () => blacklistRegion(gene,coord,"probe"));
            });
        }
    }

    function sendDataToBackend(gene, coord, region, status) {
        // The data to send
        const sampleAssay = {{ smp_grp | tojson | safe }}
        const payload = { gene: gene, region: region, coord: coord, status: status, smp_grp: sampleAssay };

        // Send the data to the backend using fetch
        fetch('/update-gene-status', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(payload),
        })
        .then(response => response.json())
        .then(data => {
            alert('Success:' + data.message);
        })
        .catch(error => {
            console.error('Error:', error);
        });
    }

    function blacklistRegion(gene,coord,region) {
        // Example of sending true/false
        sendDataToBackend(gene, coord, region, "blacklist");
    }
    document.addEventListener("DOMContentLoaded", function() {
    const table = document.getElementById("sortable-table");
    const headers = table.querySelectorAll("th");

    headers.forEach(header => {
        header.addEventListener("click", function() {
            const columnIndex = this.getAttribute("data-column");
            const type = this.getAttribute("data-type");
            const tbody = table.querySelector("tbody");
            const rows = Array.from(tbody.querySelectorAll("tr"));

            // Determine sort direction
            const isAscending = this.getAttribute("data-order") === "asc";
            this.setAttribute("data-order", isAscending ? "desc" : "asc");

            // Sort function
            const sortedRows = rows.sort((a, b) => {
                let aValue = a.children[columnIndex].textContent.trim();
                let bValue = b.children[columnIndex].textContent.trim();

                if (type === "number") {
                    aValue = parseFloat(aValue) || 0;
                    bValue = parseFloat(bValue) || 0;
                }

                return isAscending ? aValue > bValue ? 1 : -1 : aValue < bValue ? 1 : -1;
            });

            // Append sorted rows
            tbody.innerHTML = "";
            sortedRows.forEach(row => tbody.appendChild(row));
        });
    });
});
</script>

{% endblock %}