{% extends "report_layout.html" %}
{% block body %}

{% set germline = True if sample.name == "11400-19" %}

{% if save != 1 %}
  <div class="report-preview-card">
    <h2 class="report-heading">*** PREVIEW OF REPORT ***</h2>

    {% if has_access('create_report', min_role='admin') %}
      <a href="{{ url_for('dna_bp.save_dna_report', sample_id=sample.name) }}" class="report-button">Finalize Report</a>
    {% endif %}
  </div>
{% endif %}


<!--  Report Header -->
<span class="report_header">{{ assay_config.reporting.report_header }} - {{ sample.name }}</span>

<!--  Report Patient/Analysis Information -->
<div class="report_div">
  <table class='report_general'>
    <tr>
      <td class="top_report_key">Patientnamn</td>
      <td class="top_report_val">&lt;PATIENT_NAME&gt;</td>
    </tr>
    <tr>
      <td class="top_report_key">Personnummer</td>
      <td class="top_report_val">&lt;PERSONAL_IDENTITY_NUMBER&gt;</td>
    </tr>
    <tr>
      <td class="top_report_key">Prov-ID</td>
      <td class="top_report_val">{{ sample.name }}
    </tr>
    {% if sample.num_samples == 2 %}
      <tr>
        <td class="top_report_key">KontrollProv-ID</td>
        <td class="top_report_val">{{ sample.control_id }}
      </tr>
    {% endif %}
    <tr>
      <td class="top_report_key">Registreringsdatum</td>
      <td class="top_report_val">&lt;REGISTRATION_DATE&gt;</td>
    </tr>
    <tr>
        <td class="top_report_key">Provtyp</td>
        <td class="top_report_val">&lt;TUMOR_SAMPLE_TYPE&gt;</td>
    </tr>
    {% if assay_group not in ["swea", "gmsonco"] %}
    <tr>
      <td class="top_report_key">Frågeställning</td>
      <td class="top_report_val">
        {% if assay_group == "myeloid" %}
          {% if sample.num_samples == 2 %}
            Hematologisk neoplasi
          {% else %}
            &lt;DIAGNOSIS&gt;
          {% endif %}
        {% elif assay_group == "solid" %}
          {% if 'subpanel' in sample %}
            {% if sample.subpanel == "BP" %}
              Bröst-Pilot
            {% else %}
              {{sample.subpanel}}
            {% endif %}
          {% endif %}
        {% endif %}
      </td>
    </tr>
    {% endif %}
    <tr>
      <td class="top_report_key">Rapportdatum</td>
      <td class="top_report_val">{{ report_date }}</td>
    </tr>
    <tr>
      <td class="top_report_key">Analysmetod</td>
      <td class="top_report_val">{{ assay_config.reporting.report_method }}</td>
    </tr>
    <tr>
      <td class="top_report_key">Analys genomförd av</td>
      <td class="top_report_val">
          Centrum för molekylär diagnostik (CMD) och Klinisk genetik och patologi
      </td>
    </tr>
    <tr>
      <td class="top_report_key">Rapport genererad av</td>
      <td class="top_report_val">{{ current_user.fullname }}</td>
    </tr>
    <tr>
      <td class="top_report_key">Rapport-ID</td>
      <td class="top_report_val">
        {% if sample.report_num is defined %}
          {{ sample.name }}.{{ sample.report_num + 1 }}
        {% else %}
          {{ sample.name }}.1
        {% endif %}
      </td>
    </tr>
  </table>
</div>

<!--  Variant Summary -->
<span class="report_header">Analysresultat</span>
{% if "SNV" in report_sections %}
  {% set variants = report_sections_data.snvs %}
  <span class="report_header">Kliniskt relevanta SNVs och små INDELs</span>
  <table class="variant_table">
    <tr>
      <th>Gen</th>
      <th>Variant</th>
      <th>Exon/Intron</th>
      <th>Variantfrekvens</th>
      <th>Klassificering</th>
    </tr>
    {% for var in variants %}
        <tr>
          <td>{{ var.symbol }}</td>
          <td>{{ var.variant|unesc }}</td>
          <td>
            {% if var.exon %}
              E{{ var.exon[0] }} of {{ var.exon[1] }}
            {% elif var.intron %}
              I{{ var.intron[0] }} of {{ var.intron[1] }}
            {% else %}
              -
            {% endif %}
          </td>
          <td>{{ var.af|perc_no_dec }}</td>
          <td>{{ var.class_short_desc }}</td>
        </tr>
    {% else %}
        <tr>
          <td>Inga detekterade varianter</td>
        </tr>
    {% endfor %}
  </table>
{% endif %}

<!-- CNV Summary -->
{% if "CNV" in report_sections %}
  {% set cnvs = report_sections_data.cnvs %}
  <span class="report_header">Kliniskt relevanta kopietalsförändringar</span>
  <table class="variant_table">
    <tr>
      <th>Gen(er)</th>
      <th>Storlek</th>
      <th>Typ</th>
      <th>Kopietal</th>
    </tr>
      {% for cnv in cnvs %}
        <tr>
          <td>
            {% set non_panel_genes = [0] %}
            {% for gene in cnv.genes%}
              {% if gene.class %}
                {{gene.gene}}
              {% else %}
                {% if non_panel_genes.append(non_panel_genes.pop()+1) %}{% endif %}
              {% endif %}
            {% endfor %}
            {% if non_panel_genes[0] > 0 %}
              <font color='#888'>+ {{non_panel_genes[0] }} other genes</font>
            {% endif %}
          </td>
          <td>{{ cnv.size }} bp</td>
          <td>
            {% if cnv.ratio > 1 %}
              AMP
            {% elif cnv.ratio > 0 %}
              DUP
            {% else %}
              DEL
            {% endif %}
          </td>
          <td>{{ 2*(2**cnv.ratio)|round(2) }}</td>
        </tr>
      {% else %}
        <tr>
          <td>
            Inga detekterade kopietalsförändringar.
          </td>
        </tr>
      {% endfor %}
  </table>
{% endif %}

<!-- DNA Fusions Summary -->
{% if "TRANSLOCATION" in report_sections %}
  {% set transloc = report_sections_data.translocs %}
  <span class="report_header">Kliniskt relevanta fusioner (DNA)</span>
  <table class="variant_table">
    <tr>
      <th>Gen 1</th>
      <th>Gen 2</th>
      <th>HGVS.p</th>
    </tr>
    {% for tl in translocs %}
      {% set sel_ann = tl.INFO.MANE_ANN or tl.INFO.ANN[0] %}
      {% set genes = sel_ann.Gene_Name.split('&') %}
      <tr>
        <td>{{ genes[0] }}</td>
        <td>{{ genes[1] }}</td>
        <td>{{ sel_ann.HGVSp|unesc }}</td>
      </tr>
    {% else %}
      <tr>
        <td>Inga detekterade fusioner</td>
      </tr>
    {% endfor %}
  </table>
{% endif %}

<!-- Biomarkers Summary -->
{% if 1==0 %}
{% if "BIOMARKER" in report_sections %}
  {% set biomarkers = report_sections_data.biomarkers %}
  <span class="report_header">Andra biomarkörer</span>
  <table class="variant_table">
    <tr>
      <th>Markör</th>
      <th>Värde</th>
    </tr>
    {% for bio in biomarkers %}
      {% if "MSIS" in bio %}
        <tr>
          <td>
            MSI(Single)
          </td>
          <td>
            {{bio.MSIS.perc}}%
          </td>
        </tr>
      {% endif %}
      {% if "MSIP" in bio %}
        <tr>
          <td>
            MSI(Paired)
          </td>
          <td>
            {{bio.MSIP.perc}}%
          </td>
        </tr>
      {% endif %}
      {% if "HRD" in bio %}
        <tr>
          <td>
            HRD
          </td>
          <td>
            {{bio.HRD.sum}}
          </td>
        </tr>
      {% endif %}
    {% else %}
      <tr>
        <td>
          Inga biomarkörer uppmätta
        </td>
        <td>
          -
        </td>
      </tr>
    {% endfor %}
  </table>
{% endif %}
{% endif %}

<!--  Report Comments/Result Summary -->
<span class="report_header">Slutsats</span>
<div class="conclusion">
  <div class="results_summary">
    {% if sample.comments|length > 0 %}
      {% for comment in sample.comments if comment.hidden != 1 %}
        {% if loop.last %}
          {{ comment.text|markdown|safe }}
        {% endif %}
      {% endfor %}
    {% else %}
      Slutsats saknas!
    {% endif %}
  </div>
</div>

<!-- Page break for the detailed results -->
<p style="page-break-before: always;"></p>
<br>

<!-- CNV Profile for WGS samples -->
{% if report_sections_data.get('cnv_profile_base64') %}
  <span class="report_header">Kopietalsprofil, genomisk översikt</span>
  <div style="transform:rotate(90deg);">
    <img style="width:900px; height:auto; padding:30px;" src="data:image/png;base64,{{ report_sections_data.cnv_profile_base64 }}">
  </div>
<p style="page-break-after: always;"></p>
{% endif %}

<!-- Detailed list of each variants reported -->
{% if variants %}
  <span class="report_header">Detekterade varianter</span>
  {% for var in variants %}
    <div class=report_div style="page-break-inside:avoid;">
      <table class='report_variant'>
        <tr>
          <th class="report_variant_header" colspan=6>
            {% if var.variant %}
              <span class="report_variant_header">{{ var.symbol }}: {{ var.variant|unesc }}</span><br>
            {% else %}
            {#{ var.INFO }#}
            {% endif %}
          </th>
        </tr>
        <tr>
          <td class="report_key">Gen</td>
          <td class="var_report_val">{{ var.symbol }}</td>

          <td class="report_key">Typ</td>
          <td class="var_report_val">{{var.class_type}} {{ var.variant_class }}</td>

          <td class="report_key">Transkript</td>
          <td class="var_report_val">
            {{ var.feature }}
            {% if var.exon %}
              , exon {{ var.exon[0] }}
            {% elif var.intron %}
              , intron {{ var.intron[0] }}
            {% endif %}
          </td>
        </tr>
        <tr>
          <td class="report_key">Konsekvens</td>
          <td class="var_report_val">{{ var.consequence }}</td>
          <td class="report_key">cDNA-förändring</td>
          {% if var_type == "snv" %}
            <td class="var_report_val">
              <div class="wrapping">{{ var.cdna|unesc }}</div>
            </td>
          {% else %}
            <td class="var_report_val">{{ var.cdna }}</td>
          {% endif %}

          <td class="report_key">Proteinförändring</td>
          <td class="var_report_val">
            {% if var.protein_changes %}
                {% for p_change in var.protein_changes %}
                  {{ p_change|safe }}<br>
                {% endfor %}
              {% else %}
            {% endif %}
          </td>
        </tr>
        <tr>
          <td class="report_key">Kromosom</td>
          <td class="var_report_val">chr{{ var.chr }}</td>

          <td class="report_key">Position (hg38)</td>
          <td class="var_report_val">{{ var.pos }}</td>

          <td class="report_key">Frekvens</td>
          <td class="var_report_val">{{ var.af|perc_no_dec }}</td>
        </tr>

        {% if var.class != 999 %}   
          <tr>
            <td class="report_key">Klassificering</td>
            <td class="report_annotation" colspan=5>
              {{ var.class|format_tier }} -
              {{ var.class_long_desc }}<br>
            </td>
          </tr>
        {% endif %}

        {% if var.global_annotations|length > 0 %}
          <tr>
            <td class="report_key">Kommentar</td>
            {% if var.annotations_interesting|length > 0 %}
              {% for assay_sub, key in var.annotations_interesting.items() %}
                <td class="report_annotation" colspan=5>
                  {{ key.text|format_comment|safe }}
                </td>
              {% endfor %}
            {% else %}
              {% set sorted_annos = var.global_annotations|sort(attribute="time_created", reverse=True) %}
              <td class="report_annotation" colspan=5>
                {{ sorted_annos[0].text|format_comment|safe }}
              </td>
            {% endif %}
          </tr>
        {% endif %}

        {% if var.comments|length > 0 and 1==2 %} {# DO NOT SHOW COMMENTS EVER #}
        <tr>
          <td class="report_key">
            Kommentar
          </td>
          <td class="report_comment" colspan=7>
            {% for comment in var.comments %}
              <li>{{ comment.text|format_comment|safe }}
            {% endfor %}
          </td>
        </tr>
        {% endif %}
      </table>
    </div>
  {% endfor %}
{% endif %}

<!-- Analysis Description -->
<span class="report_header">Analysbeskrivning</span>
<div class="analysis_description">
  {{ assay_config.reporting.report_description|safe }}

  <form id="genelist-form" action="{{ url_for('common_bp.get_sample_genelists', sample_id=sample.name, sample_assay=sample_assay, _external=True) }}" method="POST" target="_blank">
    <input type="hidden" name="enc_genelists" id="enc-genelists-input">
    <input type="hidden" name="enc_panel_doc" id="enc-panel-input">
    <input type="hidden" name="enc_sample_filters" id="enc-sample-filters-input">
  </form>
  
  {% set ns = namespace(table_no=1) %}
  <div>
    {% for genelist_name, genelist_values in genes_covered_in_panel.items() %}
      {% if genelist_name != sample.assay %}
        <p><b>Tabell {{ ns.table_no }}: Gener inkluderade i <i>{{ genelist_name }}</i> insilico-panel</b></p>
        <div>
          <table class="genetable" id="snvs-table">
            <tbody>
              {% for i in range(0, genelist_values.covered | length, 15) %}
                <tr>
                  {% for gene in genelist_values.covered[i:i+15] %}
                    <td>{{ gene }}</td>
                  {% endfor %}
                </tr>
              {% endfor %}
            </tbody>
          </table>
        </div>
        {% set ns.table_no = ns.table_no + 1 %}
      {% endif %}
    {% endfor %}
  </div>


  <p>
    For complete panel gene lists, and configurations, see 
    <a href="#" onclick="submitGenelistForm(event)" class="text-blue-600 hover:underline" target="_blank">here.</a> 
    <br>Or, for ASP genelist, visit <a href="{{url_for("public_bp.asp_genes",asp_id=sample.assay, _external=True)}}"> {{url_for("public_bp.asp_genes",asp_id=sample.assay , _external=True)}} </a>
  </p>


  {% if not germline %}
    <p>
    <b>Tabell {{ns.table_no}}: Förklaring av klassificering</b>
    <p>
    <table class="info">
      <tr>
        <td>Tier I</td>
        <td>Variant av stark klinisk signifikans (genvarianter som är diagnostiska, behandlingsstyrande eller
          riskstratifierande enligt gällande riktlinjer)</td>
      </tr>
      <tr>
        <td>Tier II</td>
        <td>Variant av potentiell klinisk signifikans (varianter i gener beskrivna i ett flertal publikationer)</td>
      </tr>
      <tr>
        <td>Tier III</td>
        <td>Variant av oklar klinisk signifikans (varianter i gener beskrivna i ett fåtal publikationer)</td>
      </tr>
      <tr>
        <td>Tier IV</td>
        <td>Variant bedömd som benign eller sannolikt benign</td>
      </tr>
    </table>

    <p>
      <b>Referenser</b>
    <p>
      [1] Standards and guidelines for the interpretation and reporting of sequence variants in cancer. Li et al, Journal
      of Molecular Diagnostics, 2017.
  {% endif %}
</div>


<!-- Embed assay_panel_doc -->

<script type="application/json" id="encrypted-genelists">{{ encrypted_genelists }}</script>
<script type="application/json" id="encrypted-panel-doc">{{ encrypted_panel_doc }}</script>
<script type="application/json" id="encrypted-sample-filters">{{ encrypted_sample_filters }}</script>

<script>
  // Function to submit the form with encrypted genelists and panel doc
  function submitGenelistForm(e) {
    e.preventDefault();
  
    const genelistsEnc = document.getElementById("encrypted-genelists").textContent.trim();
    const panelEnc = document.getElementById("encrypted-panel-doc").textContent.trim();
    const sampleFiltersEnc = document.getElementById("encrypted-sample-filters").textContent.trim();
  
    document.getElementById("enc-genelists-input").value = genelistsEnc;
    document.getElementById("enc-panel-input").value = panelEnc;
    document.getElementById("enc-sample-filters-input").value = sampleFiltersEnc;
  
    document.getElementById("genelist-form").submit();
  }
  </script>
  

{% endblock %}